Avoid flakiness for vacuum_drop_phase_ao test

ashwinstar opened the pull request:
Add helper wait_until_waiting_for_required_lock() to check if query
has reached the intended blocking point, which is waiting for lock to
be granted. Only after this state is reached, we wish to execute next
command to unblock the same.

Currently, the isolation2 framework lacks to provide direct support
for such blocking behavior. Hence, for short term I feel adding the
common helper function is good step. In long term we can similar to
isolation framework add something simple and generic for this. Though
I like the explicit checking for exact lock type and relation
etc.. the current helpful function provides.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

ashwinstar commented:
Funny, the pipeline failed with 
```
--- \/tmp\/build\/b235ddbb\/gpdb_src\/src\/test\/isolation2\/expected\/vacuum_drop_phase_ao\.out	2019-09-09 23:20:01.825956087 +0000
+++ \/tmp\/build\/b235ddbb\/gpdb_src\/src\/test\/isolation2\/results\/vacuum_drop_phase_ao\.out	2019-09-09 23:20:01.829956087 +0000
@@ -3,7 +3,7 @@
 -- an abort) and thus guarantees that seg file states are consistent across QD/QE.
 
 include: helpers/server_helpers.sql;
-CREATE
+ERROR:  tuple concurrently updated (heapam.c:4631)
```
Two parallel tests seems tried to create or replace something and hence encountered the error. Means we can't have tests with includes in same parallel group, that's very unfortunate and needs to be resolved. No clear ideas yet on how to do it. I don't like the idea of having setup file upfront as that adds additional dependency to run the individual test. No sure what can resolve the issue.

ashwinstar commented:
After thinking more and discussing with @d as well, seems upfront `setup` test way is only good way out too. That resolves helps to avoid code duplication plus, also helps to only one time create these objects and save on execution time (little bit).

We talked little bit to resolve the situation using different schemas in different tests, but didn't like that solution. Plus, also idea was to create extension instead of `setup` and load the same similar to `gp_inject_fault` but even for that one needs to do install the sql unnecessarily compared to that setup test seems straightforward.

So, most likely will do convert the `server_helpers.sql` into setup. Please let me know if there are better alternatives.

asimrp commented:
Thank you for posting the summary of your discussions.  Found it very helpful.  I also think having a setup step that is executed once prior to running a bunch of tests concurrently is the way to go.

What flaky error is this PR trying to address?  It will be good to mention that somewhere.

paul-guo- commented:
+1 for the setup solution and extracting the common function.

soumyadeep2007 commented:
@asimrp This is the flake:

```
--- \/tmp\/build\/e18b2f02\/gpdb_src\/src\/test\/isolation2\/expected\/vacuum_drop_phase_ao\.out    2019-09-05 00:14:41.580197372 +0000
+++ \/tmp\/build\/e18b2f02\/gpdb_src\/src\/test\/isolation2\/results\/vacuum_drop_phase_ao\.out    2019-09-05 00:14:41.580197372 +0000
@@ -32,11 +32,12 @@
 DELETE 4
 -- We should see that VACUUM blocks while the QE holds the access shared lock
 1&: VACUUM ao_test_drop_phase;  <waiting ...>
+FAILED:  Forked command is not blocking; got output: VACUUM
 
 0U: END;
 END
 1<:  <... completed>
-VACUUM
+FAILED:  Execution failed
 
 -- We should see that the one visible tuple left after the DELETE gets compacted
 -- from segno = 1 to segno = 2.

======================================================================
```

ashwinstar commented:
Okay, so I moved the function to `setup` test and using the same in the two tests. I didn't move the whole `server_helpers.sql` to `setup` yet. We can see how we feel about it for some more time and decide on it little later.

Also, @asimrp I updated commit message to explain what flakiness the commit it trying to address. Thanks for popping the question to help add the same explicitly.
